Shopify/cli
# Enable external access to cli inside a spin instance

### WHY are these changes introduced?
part of  https://github.com/Shopify/shopify-cli-planning/issues/426
As part of the effort in [streamlining the first-party experience when creating apps](https://vault.shopify.io/gsd/projects/25765), we would like to make it as easy as possible for 1p developers to create apps on Spin.

Despite the environment used to run the `dev` command the CLI proceses must be accesible from the browser. Right now, the cli is not accessible when is run inside a spin instance. 

### WHAT is this pull request doing?
- Adds a `constellation` configuration to the CLI repo thus `1P users` can create a spin instance with the services needed to run CLI commands. Besides the port needed for the CLI to initiate the proceses for authentication, dev, etc, ... is opened in the `nginx` configuration. [Spin doc](https://vault.shopify.io/page/Preparing-repositories~dhbc579.md)

- In the same way that with the other cloud environments, the cli processes are configured using a public `spin domain url`  using the above configuration that allows external access to cli

- Modified `dev` [internal flowchart ](https://www.figma.com/file/PCukduZbTX4IWhC8mkGpHO/CLI3-Internals-(Copy)?node-id=0%3A1)

### How to test your changes?
- Checkout the branch used to create this PR `feature/add_external_access_to_cli_inside_spin`
- Create an spin instance using this command `spin up path/to/branch/.spin/constellation.yml -c cli.branch=feature/add_external_access_to_cli_inside_spin`
- Access to the spin instance `spin shell`
- Please make sure all the services are up and running. It's possible that some service is failing in case you receive `502` errors when opening the dashboard or using the apis through the CLI,
- A new development store should be created. In case the store is not listed once created, restart the partner service to see it immediately (it should be listed after waiting sometime. There is an issue with partners kafka consumers)
- Create a new app.
  - `cd cli`
  - `yarn install`
  - `yarn create-app --path path/to/cli-apps`
- Run `yarn shopify app dev`
- Finish the authentication process

- Preview is not working because the issue with the spin stores domains. Solved in this [PR](https://github.com/Shopify/cli/pull/553)
  
### Post-release steps0
If the users with older `partner` constellations updates the CLI version, the won't be able to run the CLI in spin, because in the new version, a `nginx` routing configuration is used instead of `ngrok` by default. The are two alternatives:
- Create a new spin instance that includes `partners` which will have the `nginx` configuration built-in
- Creating a manual tunnel inside spin and running the `dev` command with the flag `--tunell-url`

### Measuring impact

How do we know this change was effective? Please choose one:


### Checklist
